Resilience bug fixes

jtfmumm opened the pull request:
Potential bug fixes for #3017,#3025,#3026,#3027,#3028,#3030

slfritchie commented:
Yup, this is a problem for the ConnectorSink.  See logs at http://wallaroolabs-dev.s3.amazonaws.com/logs/run_aloc_sink_test.out.20191011.gz.  The relevant lines are below.

* The ConnectorSink votes to abort 2PC phase 1.
* ConnectorSink tells CheckpointInitiator to abort.
* CheckpointInitiator ignores the abort request.
* The checkpoint barrier protocol hangs, which times out the rest of the test.

```
1570812570.196347,Initiating barrier protocol for CheckpointBarrierToken(3).
1570812570.196919,DEBUG,ConnectorSink,Receive barrier CheckpointBarrierToken(3) at ConnectorSink 89082929258121052276912048928181045818
...
1570812570.197084,DEBUG,ConnectorSink,Barrier CheckpointBarrierToken(3) complete at ConnectorSink 89082929258121052276912048928181045818
1570812570.197092,DEBUG,2PC,2PC: Checkpoint state 3 at ConnectorSink 89082929258121052276912048928181045818, txn-id 
1570812570.197141,DEBUG,2PC,2PC: sent phase 1 for txn_id worker-initializer-id-89082929258121052276912048928181045818:c_id=3
1570812570.197193,Checkpoint_Initiator: Event Log CheckpointId 3 complete for worker initializer
2019-10-11 16:49:30,197 INFO     [aloc_sink:333] 2PC: abort=True next transaction: local txn count 3
1570812570.199267,DEBUG,2PC,2PC: GOT MessageMsg
1570812570.199281,DEBUG,2PC,2PC: reply for txn_id worker-initializer-id-89082929258121052276912048928181045818:c_id=3 was false
1570812570.199286,DEBUG,2PC,2PC: set 2PC state => 4
1570812570.199291,DEBUG,2PC,2PC: abort_decision: txn_id worker-initializer-id-89082929258121052276912048928181045818:c_id=3 phase 1 ABORT
1570812570.199295,DEBUG,2PC,2PC: set 2PC state => 4
1570812570.199333,Barrier CheckpointBarrierToken(3) aborted by worker initializer!
1570812570.199373,UNEXPECTED CALL to abort_checkpoint() on checkpoint initiator phase _CheckpointingPhase. Ignoring!
terminating sink
terminating wallaroo
Waiting one second for processes to exit cleanly
1570812589.474786,DEBUG,2PC,2PC: DBG: hard_close: state = 4, txn_id_at_close = worker-initializer-id-89082929258121052276912048928181045818:c_id=3, barrier_token_at_close = CheckpointBarrierToken(3)
1570812589.474818,INFO,ConnectorSink,ConnectorSink connection closed, throttling
1570812589.474823,INFO,ConnectorSink,ConnectorSink is experiencing back pressure, connected = false
1570812589.474828,INFO,ConnectorSink,RE-Connecting ConnectorSink to 127.0.0.1:7200
1570812589.474931,EventLog: dispose
```

slfritchie commented:
TODO item: The script & README changes of this branch vs. aloc-tee-source+scripts+autoscale-checkpoint need a merge.  I can do it, if this PR goes forward.

slfritchie commented:
TODO item: use `Log` module to log unexpected calls at a level above `INFO`.

slfritchie commented:
@jtfmumm: I updated comment https://github.com/WallarooLabs/wallaroo/pull/3033#issuecomment-541171818 with more info.

I think that commit 0bc9dc7's change to `abort_checkpoint()` needs to be reverted.

slfritchie commented:
Using a base of aloc-tee-source+scripts+autoscale-checkpoint branch at commit a107c1c3b, then applying this PR's commits of: 8021e15 + 0bc9dc7 + b9a991d.  Then using `./master-crasher.sh 4 crash2 crash3 crash-sink` I hit a novel crash in `worker3`:

```
Invariant violated in /home/vagrant/wallaroo/lib/wallaroo/core/recovery/event_log_phase.pony at line 330
```

Full logs at http://wallaroolabs-dev.s3.amazonaws.com/logs/logs.20191009.1570826546.tar.gz

slfritchie commented:
An update to https://github.com/WallarooLabs/wallaroo/pull/3033#issuecomment-541218917 about the "event_log_phase.pony at line 330" crash: see http://wallaroolabs-dev.s3.amazonaws.com/logs/logs.20191009.1570834161.tar.gz. With a bit of extra printf spam, I see that the set that `ack_rollback()` is querying is size 0.

```
1570834078.297075,ERROR: resilient_id = 305844631369707547586773589046351463265, _resilients_acked size = 0
Invariant violated in /home/vagrant/wallaroo/lib/wallaroo/core/recovery/event_log_phase.pony at line 336
```

slfritchie commented:
TODO maybe?  For the new `_unexpected_call()` calls, use `__loc.method_name()` instead of individual strings per function?

jtfmumm commented:
@slfritchie You're right about the abort_checkpoint change.  I've fixed that and updated to use __loc.method_name().  I've removed the script and passthrough related changes and created a new PR #3035.
